src/digest.rs

```diff
@@ -0,0 +1,119 @@
+use ff::PrimeField;
+use serde::Serialize;
+use sha3::{Digest, Sha3_256};
+use std::iter::Extend;
+use std::marker::PhantomData;
+
+use crate::constants::NUM_HASH_BITS;
+
+/// For building digests
+#[derive(Clone)]
+pub struct DigestBuilder<F: PrimeField, T: HasDigest<F>> {
+  bytes: Vec<u8>,
+  hasher: Option<Sha3_256>,
```

Comment:
What is the reasoning behind this `Option<_>`?

Reply:
I believe it was because I could not finalize the hasher while it was still owned by the builder. The option lets me `take()` it before finalization.